### ISSUE TITLE
temperature 0 is run-to-run nondeterministic on the R8 runtime — present with DFlash disabled, so not a speculation defect

### ISSUE THREAD

**Issue by sethforprivacy** [FujitsuPolycom/sparkring#161]
### Summary

`top_p` and `top_k` sampling filters are silently ignored when serving with DFlash speculative decoding. Requests with `temperature: 1` plus `top_p: 0.0001` (or `top_k: 1`) — which should decode effectively greedily — produce different outputs on every run. Plain greedy (`temperature: 0`) is byte-for-byte reproducible on the same deployment, so this is not numerical jitter: the truncation filters never reach the sampler.

Found via [llmprobe](https://github.com/ddalcu/llmprobe), which flags this as a MUST-level OpenAI-conformance failure on both the `chat/completions` and `responses` surfaces; reproduced by hand (below).

### Environment

- Image: `ghcr.io/fujitsupolycom/sparkring-glm53-sparkcache@sha256:380283a506aeb8f9d486a3c64cd738e44268c3cc21590913ea9e4685869f256a`
  (embedded revisions: sparkring `1f02801e61893e8060668948cdaed81156b3f371`, vllm `55969c16d4da57da76ee5729f3102d4b2003833c`, b12x `6255090a03b12c3f7d552102a02fac0b542fb8c9`)
- 4× DGX Spark (GB10), switchless ring, TP4 / DCP4
- Target: GLM-5.3-Flash NVFP4 (`local-inference-lab/GLM-5.3-Flash-NVFP4` @ `520de24e`); draft: `incoai/GLM-5.3-Flash-DFlash2` @ `dc77ff1c`
- Speculative config: `{"method":"dflash","num_speculative_tokens":7,"draft_tensor_parallel_size":4,"kv_cache_dtype":"auto","draft_sample_method":"probabilistic","rejection_sample_method":"standard"}`
- Target KV: fp8, `--max-num-batched-tokens 8192`, `--enable-prefix-caching`
- Model `generation_config.json` is all-null (no server-side sampling overrides)

### Repro

```bash
BODY='{"model":"GLM-5.3-Flash-Ring","temperature":1.0,"top_p":0.0001,"max_tokens":120,
  "messages":[{"role":"user","content":"Invent a two-word band name and one sentence about their sound."}]}'
curl -s $HOST/v1/chat/completions -H "Authorization: Bearer $KEY" -d "$BODY"   # run 1
curl -s $HOST/v1/chat/completions -H "Authorization: Bearer $KEY" -d "$BODY"   # run 2
```

Observed: the two completions diverge within the first few tokens (e.g. `"Simple creative task.</think>…"` vs `"Simple creative task. Invent a two-word…"`). Same with `top_k: 1` in place of `top_p` (two runs: 177 vs 181 chars, different content). With `top_p ≈ 0` or `top_k = 1` the distribution should collapse to argmax and runs should be identical.

Control: `temperature: 0` with the same prompt reproduces byte-for-byte across runs (llmprobe fidelity: 3/3 greedy prompts exact, 36/36 logprob-consistency items where the emitted token equals argmax) — so the greedy path and the logits themselves are stable.

### Suspected cause

The DFlash rejection-sampling path (`draft_sample_method: "probabilistic"`, `rejection_sample_method: "standard"`) appears to compute the draft-acceptance/bonus-token distribution from the raw temperature-scaled target probabilities without applying the request's top-p/top-k truncation, so requests sample from the unfiltered distribution regardless of the filter parameters.

Possibly related: llmprobe's same-`seed` determinism check also fails on this deployment (same seed, `temperature: 1`, different text), though a short 2×40-token idle repro of seed alone passed — the top_p/top_k symptom is the deterministic one.

### Expected

`top_p` / `top_k` (and `seed`) honored under speculative decoding, matching vLLM semantics — or, if the DFlash sampler intentionally does not support truncation filters, the request should be rejected or spec decode bypassed for it rather than silently sampling from the unfiltered distribution.

Happy to run diagnostics or test patches on this deployment — it reproduces 100% of the time.

**Comment by sethforprivacy** [comment]
Ran the speculation-off control — and it **falsified my original framing**. Correcting the record with the fuller dataset; the symptom is real but the mechanism is different from what the issue text claims.

**Control setup:** same deployment/image, idle ring (`num_requests_running` 0 throughout, single sequential requests), one prompt for every condition. Ran with `--speculative-config` omitted, then restored the normal DFlash profile and re-ran there.

**Key result — temperature 0 is itself run-to-run nondeterministic, on both paths:**

| condition (same prompt, idle, sequential) | unique outputs | first-divergence chars vs run 1 |
|---|---|---|
| temp=0, DFlash **on** | 5/5 | 21, 21, 31, 21 |
| temp=0, DFlash **off** | diverged (4 runs) | 31, 31, 31 |
| temp=1 + top_p=0.0001, DFlash on | 5/5 | 21, 40, 21, 96 |
| temp=1 + top_p=0.0001, DFlash off | diverged | 31, 56, 88 |
| temp=1 free (control), DFlash on | 3/3 | 0, 10 |

Two things fall out:

1. **`top_p`/`top_k` are most likely applied after all.** Under `top_p≈0` the runs track greedy's divergence profile exactly — long identical prefixes, flips only at the same near-tie positions where greedy itself flips — categorically unlike free temp-1 sampling (divergence at chars 0–10). My original "greedy is byte-exact, so it can't be jitter" control was run on *different prompts* (llmprobe's fidelity prompts + trivia) that happen to contain no near-ties; on this prompt greedy diverges 5/5. I withdraw the "filters never reach the sampler" claim.

2. **The actual defect: exact-tie/near-tie argmax instability at temperature 0.** Single sequential requests on an idle ring produce different greedy outputs run to run (spec on **and** off — so not DFlash-specific). A greedy run with `logprobs` shows a literal 0.0-nat top-2 gap at a flip site (`' h'` vs `' dream'`), i.e. the decode path's reduction/tie-breaking is not run-to-run stable on this stack (TP4/DCP4, B12X kernels), and any tie or near-tie token can flip. This breaks the usual "temperature 0 is reproducible for a fixed batch shape" property, and it is what llmprobe's top_p/seed MUST checks were actually detecting (those assertions presume greedy determinism).

Suggest re-titling/re-scoping this issue to the temp-0 run-to-run nondeterminism (exact-tie argmax instability) unless you'd rather I close it and file fresh. Full per-run data available; the deployment reproduces it on demand.

**Comment by sethforprivacy** [comment]
Following up on my own correction above: I have re-scoped the title to match what the control actually showed, so the issue is not sitting under a claim I withdrew.

New title: **temperature 0 is run-to-run nondeterministic on the R8 runtime — present with DFlash disabled, so not a speculation defect**

The remaining, reproducible finding, in one table (same prompt, verified-idle ring, single sequential requests):

| condition | unique outputs | first divergence (chars) |
|---|--:|---|
| temp=0, DFlash on | 5/5 | 21, 21, 31, 21 |
| temp=0, DFlash off | diverged | 31, 31, 31 |

A greedy run with `logprobs` shows a literal **0.0-nat top-2 gap** at a flip site, so the decode path's reduction/tie-breaking is not run-to-run stable on this stack (TP4/DCP4, B12X) and any exact-tie token can flip even for a single request on an idle ring. `top_p`/`top_k` are most likely fine — that part of the original report is withdrawn.

Two housekeeping notes:

- Reproducing the DFlash-off control needed a local patch to the launcher's `dflash`-only gate; a `disabled` diagnostic value would make this control reproducible without one. Raised as a one-liner in #164 rather than separately.
- Everything above still reproduces on the R10 operator image `bc7d079f…`.

If you would rather this were refiled narrowly as a fresh issue against the decode path, I am entirely fine with you closing this one — say the word and I will open the replacement. Otherwise the re-scoped title should be accurate as it stands.

**Comment by FujitsuPolycom** [comment]
I have some notes and research on this I will try to get added at some point soon, but I can confirm this is reproducible behavior on my end. 

I do wonder if its a behavior worth chasing or if it being outside the model cards recommended running params (temp 1.0 versus 0.0) leaves it in kind of a grey area. Interesting one.

**Comment by FujitsuPolycom** [comment]
Reproduced on 2026-09-04 against the published GLM-5.3 TP4/DCP4 artifact:

- image: `ghcr.io/fujitsupolycom/sparkring-glm53-sparkcache@sha256:0d4029b3b7023cf32c37ac20279469c9a2ee16a057f25aae3bcfee9ee5fb660f`
- image ID: `sha256:5e32aaa1bbe3559e81db7706ed4286248f18d27cfdb186f6b851bf786eb43075`
- vLLM composition: `e02b174693e13859de61811b5e8cd13d5308e259`
- B12X: `9ae41c5cb9935d740456479954b0089f80bd2ef2`
- topology: four GB10 ranks, TP4/DCP4, B12X KDA, embedded dual-rail SIRCL, patched NCCL fallback, DFlash2 depth seven, FP8 KV, read-write SparkCache

After the service reported zero running requests, zero waiting requests, and zero GPU KV use, five sequential Chat Completions requests used the same prompt, `temperature: 0`, and `max_tokens: 120`. Every response ended at the 120-token reasoning limit. The five complete reasoning/output byte strings produced five different SHA-256 values:

- `69a9c407e75242258284b15b06a695469fe744d5630f2d550a3aaf346fe67497`
- `015a0429a3c942d2771f91b8185f92915a667021f2b3178a0102460c763e5ebf`
- `b4fc55427739747189b6abef03d2f8c81e016599c55267f90f063babaffcd330`
- `c976bf274a193757327135bf4119390b0bc52849401555b7d36db63ffb1a4b44`
- `40e6faf3e1093a6a53376a6dd185d921ab41bd301f1269121a8a063ffe0bc30d`

Request times were 2.638–3.405 seconds. The service returned to zero running requests, zero waiting requests, and zero GPU KV use, with no CUDA, NCCL, or Python error in the observation window.

Conclusion: temperature-zero output remains nondeterministic in the published source composition. This observation does not identify DFlash, B12X, DCP, SIRCL, or SparkCache as the cause because those components were not varied. The issue remains a research diagnosis item rather than a blocker for the documented temperature-one profile. A causal test still requires matched DFlash-disabled and collective/backend variants with token-level logits at the first divergence.
